feat(sdk): support volume mount in tune API

truc0 opened the pull request:
**What this PR does / why we need it**:

As discussed on #2247 , providing a clean and simple way for specifying volume mount in `tune` API of katib Python SDK will enhance develop experience.

This PR adds `storage_per_trial` argument to `KatibClient.tune()` method. 

## Design

The `storage_per_trial` argument is designed to have the following type:

```python
List[TypedDict({
    "volume": client.V1Volume,
    "mount_path": client.V1VolumeMount,
})]
```

To simplify the usage, there are some enhancements:

1. User can omit the outer list wrapper if there is only one storage to be mount. This optimization change the type of `storage_per_trial` to: `Union[TuneStoragePerTrial, List[TuneStoragePerTrial]]`
2. User can specify a `dict` instead of a `client.V1Volume` object when the storage config is simple:
    ```python
    {
         "name": "volume-name",  # Required: Name of the volume
         "type": "pvc|secret|config_map|empty_dir",  # Required: Volume type
         # Optional fields based on type:
         # For PVC:
         "claim_name": "pvc-name",
         "read_only": False,
         # For Secret:
         "secret_name": "secret-name", 
         "items": [...],
         "default_mode": 0644,
         "optional": False,
         # For ConfigMap:
         "config_map_name": "config-name",
         "items": [...],
         "default_mode": 0644,
         "optional": False,
         # For EmptyDir:
         "medium": None,
         "size_limit": None
     }
    ```
3. User can specify `mount_path` with a `str` instead of `client.V1VolumeMount`

## Example Usage

```
storage_per_trial = [
    {
        "volume": {
            "name": "data",
            "type": "pvc",
            "claim_name": "my-data-pvc"
        },
        "mount_path": "/data"
    },
    {
        "volume": {
            "name": "config",
            "type": "config_map",
            "config_map_name": "model-config"
        },
        "mount_path": "/etc/config"
    }
]
```

<details>
<summary>Full Example</summary>


```python
import kubeflow.katib as katib
from kubernetes import client
import json
import base64

VOLUME_NAME = "katib-secret-volume"
SECRET_NAME = "katib-secret-test"

# create the secret (for testing purposes, optional)
secret = client.V1Secret(
    metadata=client.V1ObjectMeta(name=SECRET_NAME),
    data={
        "credentials.json": base64.b64encode(
            json.dumps({"username": "admin", "password": "password"}).encode()
        ).decode()
    },
)
# client.CoreV1Api().create_namespaced_secret(namespace="kubeflow", body=secret)


def objective(parameters):
    import os

    result = 4 * int(parameters["x"]) - float(parameters["y"]) ** 2
    # The result will be negative if the secret is mounted successfully
    if os.path.exists("/secret/credentials.json"):
        result *= -1
    print(f"result={result}")


parameters = {
    "x": katib.search.int(min=10, max=20),
    "y": katib.search.double(min=0.1, max=0.2),
}


katib_client = katib.KatibClient(namespace="kubeflow")

storage_per_trial = {
    "volume": {
        "name": VOLUME_NAME,
        "type": "secret",
        "secret_name": SECRET_NAME,
    },
    "mount_path": "/secret",
}

name = "katib-experiment-secret"
katib_client.tune(
    name=name,
    parameters=parameters,
    objective=objective,
    objective_metric_name="result",
    max_trial_count=12,
    resources_per_trial={"cpu": "2"},
    storage_per_trial=storage_per_trial,
)

katib_client.wait_for_experiment_condition(name=name)
print(katib_client.get_optimal_hyperparameters(name=name))
```

</details>

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2247 

**Checklist:**

- [x] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing
- [x] Support mounting volume
- [x] Documenting usage of `storage_per_trial`

mahdikhashan commented:
@truc0 thank you - would you please add unit tests and e2e for your changes -

unit tests in this path: `sdk/python/v1beta1/kubeflow/katib/api/katib_client_test.py`
e2e here: `test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py`

I'll review the code and functionality after it. thanks for your time.

Electronic-Waste commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -288,6 +297,21 @@ class name in this argument.
+                - volume: Either a kubernetes.client.V1Volume object or a dictionary
+                  containing volume configuration with required fields:
+                  - name: Name of the volume
+                  - type: One of "pvc", "secret", "config_map", or "empty_dir"
```
```suggestion
                  - type: One of "pvc", "secret", "configmap", or "empty_dir"
```

Usually, we'll name it as `configmap` instead of `config_map`:)

Electronic-Waste commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -288,6 +297,21 @@ class name in this argument.
+                  - For pvc: claim_name, read_only (optional)
+                  - For secret: secret_name, items (optional), default_mode (optional),
+                    optional (optional)
+                  - For config_map: config_map_name, items (optional), default_mode
+                    (optional), optional (optional)
```
Same naming convention issue.

Electronic-Waste commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -503,6 +527,134 @@ class name in this argument.
+                                    optional=storage["volume"].get("optional", False),
+                                ),
+                            )
+                        elif volume_type == "config_map":
+                            volume = client.V1Volume(
+                                name=volume_name,
+                                config_map=client.V1ConfigMapVolumeSource(
+                                    name=storage["volume"].get("config_map_name"),
```
Same naming convention issue.

Electronic-Waste reviewed (commented):
@truc0 I'm so sorry for the late response. Here are my initial comments for you.

Please also take a look at this PR if you are available @kubeflow/wg-automl-leads @helenxie-bit @mahdikhashan @Doris-xm

Electronic-Waste commented:
/ok-to-test
/rerun-all

Electronic-Waste commented:
Could you please fix the pre-commit error? Now, you can rerun the CI test with `/rerun-all` on yourself.

mahdikhashan commented:
> @truc0 I'm so sorry for the late response. Here are my initial comments for you.
> 
> Please also take a look at this PR if you are available @kubeflow/wg-automl-leads @helenxie-bit @mahdikhashan @Doris-xm

thanks @Electronic-Waste , i already have requested some changes here: https://github.com/kubeflow/katib/pull/2508#issuecomment-2657087766

cc: @truc0

mahdikhashan commented:
@truc0 would you please check the linter issues as well? thanks.

truc0 commented:
Sorry for late reply, I have rebased the master branch and added some corresponding changes.

Thanks @Electronic-Waste  for pointing out the naming convention problem, the `config_map` should be renamed to `configmap` in the latest version.

Thanks @mahdikhashan for mentioning the unit test and the e2e test. Some unit tests about the `storage_per_trial` has been added in the latest commit. However, like the `env_per_trial` param, the `storage_per_trial` requires an existing volume to work. As a result, it is hard to be fully tested by unit tests and e2e test based on current test framework.

I need some advices for the test to `storage_per_trial`, is it available to create a volume in the e2e test so that this function can be tested? Or we just the format of the input value to the param like what `env_per_trial` do.

Electronic-Waste commented:
> I need some advices for the test to storage_per_trial, is it available to create a volume in the e2e test so that this function can be tested?

Thanks for the update @truc0! You can create secret in the cluster like issue #2247 request, and specify it in the e2e test.

FYI, you can modify the existing tune API e2e test in: https://github.com/kubeflow/katib/blob/master/test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py

mahdikhashan commented:
> Sorry for late reply, I have rebased the master branch and added some corresponding changes.
> 
> Thanks @Electronic-Waste for pointing out the naming convention problem, the `config_map` should be renamed to `configmap` in the latest version.
> 
> Thanks @mahdikhashan for mentioning the unit test and the e2e test. Some unit tests about the `storage_per_trial` has been added in the latest commit. However, like the `env_per_trial` param, the `storage_per_trial` requires an existing volume to work. As a result, it is hard to be fully tested by unit tests and e2e test based on current test framework.
> 
> I need some advices for the test to `storage_per_trial`, is it available to create a volume in the e2e test so that this function can be tested? Or we just the format of the input value to the param like what `env_per_trial` do.

thanks for your time, i'll look into e2e testing on the weekend.
